<launch>
  <!-- argument declaration -->
  <arg name = "cmd_src" default = "none" 
  description = "publish cmd_vel using the circle node: circle,
  publish cmd_vel using teleop: teleop, disable nodes that publish cmd_vel: none" />

  <arg name = "robot" default = "nusim" 
  description = "use the nusim simulator: nusim,
  use the real turtlebot3: localhost, disable additional nodes: none" />

  <arg name = "use_rviz" default = "false" 
  description = "launch rviz: true,
  don't launch rviz: false" />

  <!-- static transform publisher -->
  <node pkg = "tf2_ros" exec="static_transform_publisher" 
  args = " --frame-id nusim/world --child-frame-id odom" />

  <!-- cmd vel publisher -->
  <group if="$(eval '\'$(var cmd_src)\' == \'teleop\'')">
   <node pkg="teleop_twist_keyboard" exec="teleop_twist_keyboard"
   output="screen" launch-prefix="xterm -e" />
  </group>

  <group if="$(eval '\'$(var cmd_src)\' == \'circle\'')">
   <node pkg="nuturtle_control" exec="circle" >
    <param name="frequency" value="100.0" /> 
  </node>
  </group>

  <node
      pkg='nuturtle_control'
      exec='turtle_control' 
      args = " --remap wheel_cmd:=red/wheel_cmd --remap sensor_data:=red/sensor_data" >
      <param from="$(find-pkg-share nuturtle_description)/config/diff_params.yaml" />
  </node>
  
  <!-- simulator -->
  <group if="$(eval '\'$(var robot)\' == \'nusim\' and \'$(var use_rviz)\' == \'true\'')">
    <arg name = "rviz_config" default = "$(find-pkg-share nuturtle_control)/config/nuturtle_nusim.rviz" 
    description = "Name of the rviz configuration file relative to the project's share directory" />
    <!-- rviz will enable us to see the robot -->
    <node pkg="rviz2" exec="rviz2" args="-d $(var rviz_config)"/>
    <include file="$(find-pkg-share nuturtle_description)/launch/load_one.launch.py">
      <arg name="color" value="blue" />
      <arg name="use_rviz" value="false" />
    </include>
    <include file="$(find-pkg-share nuturtle_description)/launch/load_one.launch.py">
      <arg name="color" value="red" />
      <arg name="use_rviz" value="false" />
    </include>
    <node pkg="nusim" exec="nusim" output="screen">
      <param from="$(find-pkg-share nuturtle_description)/config/diff_params.yaml" />
      <param from="$(find-pkg-share nusim)/config/basic_world.yaml" />        
    </node>
    <node
        pkg='nuturtle_control'
        exec='odometry' >
        <param from="$(find-pkg-share nuturtle_description)/config/diff_params.yaml" />
        <param from="$(find-pkg-share nusim)/config/basic_world.yaml" />  
        <param name="body_id" value = "blue/base_footprint" />
        <param name = "wheel_left" value = "left_joint" />
        <param name = "wheel_right" value = "right_joint" />
    </node>
  </group>

</launch>
